### PR TITLE
feat(gatsby-remark-prismjs): add support for language extensions

### LINKED ISSUE #10795: gatsby-remark-prismjs: How to extend languages?
## Summary

I like to extend Prism with more tokens to allow for more granular highlighting (like VS Code's TextMate grammars). Can this be done with Gatsby?

### Basic example

```js
import Prism from 'prismjs'

Prism.languages.insertBefore('javascript', 'punctuation', {
  definition: {
    pattern: /[a-z]\w*(?=:)/i,
    lookbehind: true,
    alias: 'property',
  },
  access: {
    pattern: /(\.\s*)[a-z_$][\w$]*/i,
    lookbehind: true,
    alias: 'property',
  },
  dom: {
    pattern: /\b(?:window|document|navigator|performance|localStorage)\b/,
    alias: 'variable',
  },
  console: /\bconsole\b/,
  'class-name': {
    pattern: /\b[A-Z][A-Za-z0-9_$]+\b/,
    alias: 'variable',
  },
})
```

### PR DESCRIPTION
**Pull request by david-nossebro**
## Description

Added support to 'gatsby-remark-prismjs' plugin to make it possible to add new language definitions and extend existing languages. 

## Related Issues

Fixes #10795

### CODE REVIEW

**Comment by david-nossebro** [comment]
I was not sure if I am suppose to add entries into the changelog or how that is managed. Please let me know if I am suppose to do those entries (or if any other change is needed) :).

**Comment by pieh** [comment]
> I was not sure if I am suppose to add entries into the changelog or how that is managed. Please let me know if I am suppose to do those entries (or if any other change is needed) :).

Changelog are generated automatically based on conventional commits (that we often manually prep when merging PR). So for the most part, it would be good if PR is titled something like `feat(gatsby-remark-prismjs): add support for language extensions`

**Comment by david-nossebro** [comment]
@pieh, I am looking into that.

**Comment by pieh** [comment]
Ok - I don't want to mess your flow - I have this ready https://github.com/gatsbyjs/gatsby/compare/master...pieh:topics/support-language-extensions-with-prism-plugin?expand=1 , but won't push it here.

You would need fetch latest master, merge that in, re-run `yarn bootstrap` to install new prettier version - and then you can run `yarn format`, which should get everything in line

**Comment by david-nossebro** [comment]
Now the formatting issue is fixed. 

I have one doubt regarding the tests. I would like to reset the "Prism" dependency between the tests, but I am not sure how to do that with Jest (or if it is possible).

**Comment by david-nossebro** [comment]
This pull-request is ready for review.

**Comment by atomiks** [comment]
Is it possible to import plugins with this as well? (if that's not already possible)

e.g. JSX plugin equivalent to

```js
import 'prismjs/plugins/prism-jsx'
```

File: packages/gatsby-remark-prismjs/README.md [hunk 1]
```diff
@@ -50,6 +50,39 @@ plugins: [
             // If setting this to true, the parser won't handle and highlight inline
             // code used in markdown i.e. single backtick code like `this`.
             noInlineHighlight: false,
+            // The below setting lets you add new Prism language definitions and
```
[the remark this answers is not in the document]
Comment (david-nossebro) [reply packages/gatsby-remark-prismjs/README.md]: Thanks for the input, I did some rework. Let me know if you think the structure is better now.

**Comment by david-nossebro** [comment]
@pieh , The build is failing in Azure Pipeline, but I can´t figure out why. All tests pass locally when I do 'yarn test'. Can you help me understand what the error is?

**Comment by DSchau** [comment]
@david-nossebro it seems to be some unrelated Azure stuff; ignore for now, and sorry for the confusion!

**Comment by david-nossebro** [comment]
@DSchau , ok thanks!

**Comment by david-nossebro** [comment]
Will the build be automatically re-triggered when the Azure issue is resolved?

**Comment by david-nossebro** [comment]
I did a mistake while doing rebase... and then another one while trying to fix it with a force-push. Is there any way to repopen this pull-request or should I do a new one?

**Comment by DSchau** [comment]
@david-nossebro if you do a `git log` you may be able to checkout to a previously working commit. Otherwise feel free to do whatever you need to do to get it back working again!

**Comment by DSchau** [comment]
The approach would be something like:

- `git log`
- find commit hash that was working
- `git co <commit-hash>`
- `git branch -D topics/support-language-extensions-with-prism-plugin`
- `git push origin --delete topics/support-language-extensions-with-prism-plugin`
- `git co -b topics/support-language-extensions-with-prism-plugin`
- `git push origin topics/support-language-extensions-with-prism-plugin`

**Comment by david-nossebro** [comment]
#11932 with the same changes was added since this got closed due to my git mistake.

**Pull request closed without merging**